client/src/main/java/com/netflix/conductor/client/http/PayloadStorage.java

```diff
@@ -0,0 +1,148 @@
+     * @param uri         the location to which the object is to be uploaded
+     * @param payload     an {@link InputStream} containing the json payload which is to be uploaded
+     * @param payloadSize the size of the json payload in bytes
+     * @throws ConductorClientException if the upload fails due to an invalid path or an error from external storage
+     */
+    @Override
+    public void upload(String uri, InputStream payload, long payloadSize) {
+        HttpURLConnection connection = null;
+        try {
+            URL url = new URI(uri).toURL();
+
+            connection = (HttpURLConnection) url.openConnection();
+            connection.setDoOutput(true);
+            connection.setRequestMethod("PUT");
+
+            try (BufferedOutputStream bufferedOutputStream = new BufferedOutputStream(connection.getOutputStream())) {
+                IOUtils.copy(payload, bufferedOutputStream);
+
+                // Check the HTTP response code
+                int responseCode = connection.getResponseCode();
+                logger.debug("Upload completed with HTTP response code: {}", responseCode);
+            }
+        } catch (URISyntaxException | MalformedURLException e) {
+            String errorMsg = String.format("Invalid path specified: %s", uri);
+            logger.error(errorMsg, e);
+            throw new ConductorClientException(errorMsg, e);
+        } catch (IOException e) {
+            String errorMsg = String.format("Error uploading to path: %s", uri);
+            logger.error(errorMsg, e);
+            throw new ConductorClientException(errorMsg, e);
+        } finally {
+            if (connection != null) {
+                connection.disconnect();
+            }
+        }
+    }
+
+    /**
+     * Downloads the payload from the given uri.
+     *
+     * @param uri the location from where the object is to be downloaded
+     * @return an inputstream of the payload in the external storage
+     * @throws ConductorClientException if the download fails due to an invalid path or an error from external storage
+     */
+    @Override
+    public InputStream download(String uri) {
+        HttpURLConnection connection = null;
+        String errorMsg;
+        try {
+            URL url = new URI(uri).toURL();
+            connection = (HttpURLConnection) url.openConnection();
+            connection.setDoOutput(false);
+
+            // Check the HTTP response code
+            int responseCode = connection.getResponseCode();
+            if (responseCode == HttpURLConnection.HTTP_OK) {
+                logger.debug("Download completed with HTTP response code: {}", connection.getResponseCode());
+                return connection.getInputStream();
+            }
+            errorMsg = String.format("Unable to download. Response code: %d", responseCode);
+            logger.error(errorMsg);
+            throw new ConductorClientException(errorMsg);
+        } catch (URISyntaxException | MalformedURLException e) {
```

Comment:
can we reuse some code with the upload?

Reply:
refactoring here won't help much with readability